In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
data=pd.read_csv(r"C:\Users\sohaibkhan\Documents\Pakistan Largest Ecommerce Dataset.csv[1]\Pakistan Largest Ecommerce Dataset.csv")
data.head(3)

C:\Users\sohaibkhan\AppData\Local\Temp\ipykernel_10388\2846454908.py:1: DtypeWarning: Columns (1,2,3,7,8,9,11,12,13,14,17,18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  data=pd.read_csv(r"C:\Users\sohaibkhan\Documents\Pakistan Largest Ecommerce Dataset.csv[1]\Pakistan Largest Ecommerce Dataset.csv")


,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,sales_commission_code,...,Month,Customer Since,M-Y,FY,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,\N,...,7.0,2016-7,7-2016,FY17,1.0,NaN,NaN,NaN,NaN,NaN
1,211133.0,canceled,7/1/2016,kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...,240.0,1.0,240.0,100147444,Beauty & Grooming,\N,...,7.0,2016-7,7-2016,FY17,2.0,NaN,NaN,NaN,NaN,NaN
2,211134.0,canceled,7/1/2016,Ego_UP0017-999-MR0,2450.0,1.0,2450.0,100147445,Women's Fashion,\N,...,7.0,2016-7,7-2016,FY17,3.0,NaN,NaN,NaN,NaN,NaN


In [3]:
data.shape

(1048575, 26)

In [4]:
# Understanding of the dataset

In [5]:
# | Column Name       | Meaning                                 |
# | ----------------- | --------------------------------------- |
# | `item_id`         | Unique identifier for the item          |
# | `status`          | Order status (complete, canceled, etc.) |
# | `created_at`      | Date of order                           |
# | `sku`             | Product code                            |
# | `price`           | Product price                           |
# | `qty_ordered`     | Quantity ordered                        |
# | `grand_total`     | Total amount (likely = price × qty)     |
# | `category_name_1` | Product category                        |
# | `Customer Since`  | When the customer started buying        |
# | `FY`              | Financial year                          |
# | `Customer ID`     | Unique customer identifier              |


In [6]:
#  2. What Does “Best Price” Mean?
# You want to predict the best price of a product. This could mean:

# Option A: Optimal price to maximize sales or revenue
# We model the relationship between features like category, quantity ordered, customer, etc., and price.

# Option B: Predict a product’s expected price
# We model price as the target — to predict future or ideal prices based on product attributes.

# ✅ We'll go with Option B: Predict price based on product, category, and customer features.

In [7]:
 data.isnull().sum()

item_id                   464051
status                    464066
created_at                464051
sku                       464071
price                     464051
qty_ordered               464051
grand_total               464051
increment_id              464051
category_name_1           464215
sales_commission_code     601229
discount_amount           464051
payment_method            464051
Working Date              464051
BI Status                 464051
 MV                       464051
Year                      464051
Month                     464051
Customer Since            464062
M-Y                       464051
FY                        464051
Customer ID               464062
Unnamed: 21              1048575
Unnamed: 22              1048575
Unnamed: 23              1048575
Unnamed: 24              1048575
Unnamed: 25              1048575
dtype: int64

In [8]:
# Drop unwanted columns
data.drop(columns=["Unnamed: 21", "Unnamed: 22", "Unnamed: 23", "Unnamed: 24", "Unnamed: 25"], inplace=True, errors='ignore')






In [9]:
data.head(1)

,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,sales_commission_code,...,payment_method,Working Date,BI Status,MV,Year,Month,Customer Since,M-Y,FY,Customer ID
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,\N,...,cod,7/1/2016,#REF!,"1,950",2016.0,7.0,2016-7,7-2016,FY17,1.0


In [10]:
print(data.columns.tolist())


['item_id', 'status', 'created_at', 'sku', 'price', 'qty_ordered', 'grand_total', 'increment_id', 'category_name_1', 'sales_commission_code', 'discount_amount', 'payment_method', 'Working Date', 'BI Status', ' MV ', 'Year', 'Month', 'Customer Since', 'M-Y', 'FY', 'Customer ID']


In [11]:
data["sales_commission_code"].head(4)

0             \N
1             \N
2             \N
3    R-FSD-52352
Name: sales_commission_code, dtype: object

 | Column Name                 | Simple Explanation                                                                    | Example Values           | Data Type                             | Nature                     |
 | --------------------------- | ------------------------------------------------------------------------------------- | ------------------------ | ------------------------------------- | -------------------------- |
 | **item\_id**                | A unique number identifying the product in the order.                                 | `211131`                 | Numeric (int)                         | Categorical (ID)           |
 | **status**                  | Status of the order (e.g., completed, fraud, pending).                                | `'complete'`, `'fraud'`  | String (object)                       | Categorical                |
 | **created\_at**             | Date and time when the order was placed.                                              | `'7/1/2016'`             | String → should be `datetime`         | Time Series                |
 | **sku**                     | Stock Keeping Unit – unique code for a specific product.                              | `'kreations_YI 06-L'`    | String (object)                       | Categorical                |
 | **price**                   | Price of a single item.                                                               | `1950`                   | Numeric (float/int)                   | Numerical                  |
 | **qty\_ordered**            | Number of units ordered in that transaction.                                          | `1`, `2`                 | Numeric (int)                         | Numerical                  |
 | **grand\_total**            | Total amount of the order after taxes/discounts.                                      | `1950`                   | Numeric (float)                       | Numerical                  |
| **increment\_id**           | Unique identifier of the entire order (like an invoice ID).                           | `100147`                 | Numeric (int)                         | Categorical (ID)           |
 | **category\_name\_1**       | Product category name.                                                                | `'Women Clothing'`       | String                                | Categorical                |
 | **sales\_commission\_code** | Code used to track salesperson commission or campaign.                                | `'AFF001'`, `NaN`        | String                                | Categorical (sparse)       |
 | **discount\_amount**        | Amount of discount applied to the order.                                              | `0`, `250`               | Numeric (float)                       | Numerical                  |
 | **payment\_method**         | How the customer paid.                                                                | `'COD'`, `'Credit Card'` | String                                | Categorical                |
 | **Working Date**            | Possibly a cleaned/adjusted version of `created_at`.                                  | `'7/1/2016'`             | String → should be `datetime`         | Time Series                |
 | **BI Status**               | Business Intelligence status (unknown without more info) – maybe derived from status. | `complete`, `fraud`      | String                                | Categorical                |
 | \*\* MV \*\*                | Unknown column (check what values it holds). Might be store/brand/region.             | `NaN`, `'MV123'`         | String                                | Categorical or unknown     |
 | **Year**                    | Year of transaction.                                                                  | `2016`, `2017`           | Numeric (int)                         | Time Series                |
| **Month**                   | Month of transaction.                                                                 | `1`, `2`                 | Numeric (int)                         | Time Series                |
 | **Customer Since**          | The year the customer registered.                                                     | `2015`, `2016`           | Numeric (int)                         | Time Series                |
| **M-Y**                     | Month-Year format of transaction.                                                     | `'7-2016'`               | String → Can be converted to datetime | Time Series                |
 | **FY**                      | Financial Year. E.g., `'2016-2017'`                                                   | `'2016-2017'`            | String                                | Categorical (time-related) |
 | **Customer ID**             | Unique ID of the customer.                                                            | `C123456`                | String                                | Categorical (ID)           |


In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   item_id                584524 non-null  float64
 1   status                 584509 non-null  object 
 2   created_at             584524 non-null  object 
 3   sku                    584504 non-null  object 
 4   price                  584524 non-null  float64
 5   qty_ordered            584524 non-null  float64
 6   grand_total            584524 non-null  float64
 7   increment_id           584524 non-null  object 
 8   category_name_1        584360 non-null  object 
 9   sales_commission_code  447346 non-null  object 
 10  discount_amount        584524 non-null  float64
 11  payment_method         584524 non-null  object 
 12  Working Date           584524 non-null  object 
 13  BI Status              584524 non-null  object 
 14   MV                    584524 non-

In [14]:
# Handling the missing Values

In [15]:
data.isnull().sum()

item_id                  464051
status                   464066
created_at               464051
sku                      464071
price                    464051
qty_ordered              464051
grand_total              464051
increment_id             464051
category_name_1          464215
sales_commission_code    601229
discount_amount          464051
payment_method           464051
Working Date             464051
BI Status                464051
 MV                      464051
Year                     464051
Month                    464051
Customer Since           464062
M-Y                      464051
FY                       464051
Customer ID              464062
dtype: int64

In [16]:
data.shape

(1048575, 21)

In [17]:
# now i am droping the cloumn whcih has the maximum null values

In [18]:
data.drop(columns=["sales_commission_code"],inplace=True)

In [19]:
data.isnull().sum()

item_id            464051
status             464066
created_at         464051
sku                464071
price              464051
qty_ordered        464051
grand_total        464051
increment_id       464051
category_name_1    464215
discount_amount    464051
payment_method     464051
Working Date       464051
BI Status          464051
 MV                464051
Year               464051
Month              464051
Customer Since     464062
M-Y                464051
FY                 464051
Customer ID        464062
dtype: int64

In [20]:
data['created_at'] = pd.to_datetime(data['created_at'], errors='coerce')
data['Working Date'] = pd.to_datetime(data['Working Date'], errors='coerce')


In [21]:
data['created_at'].dt.year 

data['created_at'].dt.month 

data['created_at'].dt.dayofweek

0          4.0
1          4.0
2          4.0
3          4.0
4          4.0
          ... 
1048570    NaN
1048571    NaN
1048572    NaN
1048573    NaN
1048574    NaN
Name: created_at, Length: 1048575, dtype: float64

In [22]:
# | Code               | What It Does                                                |
# | ------------------ | ----------------------------------------------------------- |
# | `pd.to_datetime()` | Converts strings to datetime format                         |
# | `errors='coerce'`  | Replaces invalid dates with `NaT`                           |
# | Helps With         | Feature engineering (year, month, day), time-based analysis |


In [23]:
# dropping  the irrelivent columns 

In [24]:
# Drop irrelevant columns
data.drop(columns=[
    'item_id', 'sku', 'increment_id', 'Customer ID',
    'Working Date', 'sales_commission_code', 'BI Status',
    'Customer Since', 'M-Y', 'FY', ' MV ', 'status', 'created_at'
], inplace=True, errors='ignore')


In [25]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   price            584524 non-null  float64
 1   qty_ordered      584524 non-null  float64
 2   grand_total      584524 non-null  float64
 3   category_name_1  584360 non-null  object 
 4   discount_amount  584524 non-null  float64
 5   payment_method   584524 non-null  object 
 6   Year             584524 non-null  float64
 7   Month            584524 non-null  float64
dtypes: float64(6), object(2)
memory usage: 64.0+ MB


In [26]:
data.isnull().sum()

price              464051
qty_ordered        464051
grand_total        464051
category_name_1    464215
discount_amount    464051
payment_method     464051
Year               464051
Month              464051
dtype: int64

In [27]:
data.shape


(1048575, 8)

In [28]:
# | **Column**        | **Type**    | **Best Filling Strategy**                                           | **Why**                                                                         |
# | ----------------- | ----------- | ------------------------------------------------------------------- | ------------------------------------------------------------------------------- |
# | `price`           | Numeric     | ❌ **Do NOT fill with 0**<br>✅ Use **median or drop rows**           | Price = 0 ruins model. If target = price, drop rows with missing price.         |
# | `qty_ordered`     | Numeric     | ✅ Use **0 or median**                                               | 0 can mean no order. Median is better if you suspect data is missing, not zero. |
# | `grand_total`     | Numeric     | ✅ Recalculate: `price * qty_ordered`                                | Derived field — best to **compute it** rather than fill blindly.                |
# | `category_name_1` | Categorical | ✅ Use `"Unknown"` or `"Missing"`                                    | For ML models, placeholder categories work well.                                |
# | `discount_amount` | Numeric     | ✅ Fill with 0                                                       | Most likely means **no discount** was applied.                                  |
# | `payment_method`  | Categorical | ✅ Fill with `"Unknown"` or `"Missing"`                              | Keeps the row usable for classification or analysis.                            |
# | `Year`            | Numeric     | ✅ Fill with value extracted from `created_at` column (if available) | Reconstruct if possible. Else fill with 0 or drop if date matters.              |
# | `Month`           | Numeric     | ✅ Same as `Year`                                                    | Try to extract from any available datetime column.                              |


In [29]:
# Drop rows where price is missing (important for prediction)
data = data[~data['price'].isnull()]

# Fill other numerical columns
data['qty_ordered'].fillna(0, inplace=True)
data['discount_amount'].fillna(0, inplace=True)



# Fill categorical
data['category_name_1'].fillna('Unknown', inplace=True)
data['payment_method'].fillna('Unknown', inplace=True)

# Fill Year and Month with 0 or infer from datetime column if available
data['Year'].fillna(0, inplace=True)
data['Month'].fillna(0, inplace=True)


C:\Users\sohaibkhan\AppData\Local\Temp\ipykernel_10388\1536479589.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['qty_ordered'].fillna(0, inplace=True)
C:\Users\sohaibkhan\AppData\Local\Temp\ipykernel_10388\1536479589.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

Fo

In [30]:
data.isnull().sum()

price              0
qty_ordered        0
grand_total        0
category_name_1    0
discount_amount    0
payment_method     0
Year               0
Month              0
dtype: int64

In [31]:
# Save the cleaned 'data' DataFrame to a CSV file
data.to_csv('cleaned_data.csv', index=False)
